File: src/org/recast4j/detour/MeshData.java
/*
Copyright (c) 2009-2010 Mikko Mononen memon@inside.org
recast4j copyright (c) 2015-2019 Piotr Piastucki piotr@jtilia.org

This software is provided 'as-is', without any express or implied
warranty.  In no event will the authors be held liable for any damages
arising from the use of this software.
Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:
1. The origin of this software must not be misrepresented; you must not
 claim that you wrote the original software. If you use this software
 in a product, an acknowledgment in the product documentation would be
 appreciated but is not required.
2. Altered source versions must be plainly marked as such, and must not be
 misrepresented as being the original software.
3. This notice may not be removed or altered from any source distribution.
*/
package org.recast4j.detour;

import org.joml.Vector3f;

import static jdk.nashorn.internal.objects.NativeMath.max;
import static jdk.nashorn.internal.objects.NativeMath.min;
import static org.joml.Math.clamp;
import static org.recast4j.Vectors.copy;

public class MeshData {

    /**
     * The tile header.
     */
    public MeshHeader header;
    /**
     * The tile vertices. [Size: MeshHeader::vertCount]
     */
    public float[] vertices;
    /**
     * The tile polygons. [Size: MeshHeader::polyCount]
     */
    public Poly[] polygons;
    /**
     * The tile's detail sub-meshes. [Size: MeshHeader::detailMeshCount]
     */
    public PolyDetail[] detailMeshes;
    /**
     * The detail mesh's unique vertices. [(x, y, z) * MeshHeader::detailVertCount]
     */
    public float[] detailVertices;
    /**
     * The detail mesh's triangles. [(vertA, vertB, vertC) * MeshHeader::detailTriCount] See DetailTriEdgeFlags and
     * NavMesh::getDetailTriEdgeFlags.
     */
    public int[] detailTriangles;
    /**
     * The tile bounding volume nodes. [Size: MeshHeader::bvNodeCount] (Will be null if bounding volumes are disabled.)
     */
    public BVNode[] bvTree;
    /**
     * The tile off-mesh connections. [Size: MeshHeader::offMeshConCount]
     */
    public OffMeshConnection[] offMeshCons;

    public static MeshData build(NavMeshDataCreateParams params, int tileX, int tileY) {
        MeshData data = NavMeshBuilder.createNavMeshData(params);
        if (data != null) {
            data.header.x = tileX;
            data.header.y = tileY;
        }
        return data;
    }

    public static void build(MeshData data) {
        data.bvTree = new BVNode[data.header.polyCount * 2];
        data.header.bvNodeCount = data.bvTree.length == 0 ? 0
                : createBVTree(data, data.bvTree, data.header.bvQuantizationFactor);
    }

    private static int createBVTree(MeshData data, BVNode[] BVNodes, float quantFactor) {
        BVNode[] items = new BVNode[data.header.polyCount];
        for (int i = 0; i < data.header.polyCount; i++) {
            BVNode it = new BVNode();
            items[i] = it;
            it.index = i;
            Vector3f bmin = new Vector3f();
            copy(bmin, data.vertices, data.polygons[i].vertices[0] * 3);
            Vector3f bmax = new Vector3f(bmin);
            for (int j = 1; j < data.polygons[i].vertCount; j++) {
                int idx = data.polygons[i].vertices[j] * 3;
                min(bmin, data.vertices, idx);
                max(bmax, data.vertices, idx);
            }
            it.minX = clamp((int) ((bmin.x - data.header.bmin.x) * quantFactor), 0, 0x7fffffff);
            it.minY = clamp((int) ((bmin.y - data.header.bmin.y) * quantFactor), 0, 0x7fffffff);
            it.minZ = clamp((int) ((bmin.z - data.header.bmin.z) * quantFactor), 0, 0x7fffffff);
            it.maxX = clamp((int) ((bmax.x - data.header.bmin.x) * quantFactor), 0, 0x7fffffff);
            it.maxY = clamp((int) ((bmax.y - data.header.bmin.y) * quantFactor), 0, 0x7fffffff);
            it.maxZ = clamp((int) ((bmax.z - data.header.bmin.z) * quantFactor), 0, 0x7fffffff);
        }
        return NavMeshBuilder.subdivide(items, 0, data.header.polyCount, 0, BVNodes);
    }

}

File: src/org/recast4j/detour/MeshData.kt
/*
Copyright (c) 2009-2010 Mikko Mononen memon@inside.org
recast4j copyright (c) 2015-2019 Piotr Piastucki piotr@jtilia.org

This software is provided 'as-is', without any express or implied
warranty.  In no event will the authors be held liable for any damages
arising from the use of this software.
Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:
1. The origin of this software must not be misrepresented; you must not
 claim that you wrote the original software. If you use this software
 in a product, an acknowledgment in the product documentation would be
 appreciated but is not required.
2. Altered source versions must be plainly marked as such, and must not be
 misrepresented as being the original software.
3. This notice may not be removed or altered from any source distribution.
*/
package org.recast4j.detour

import jdk.nashorn.internal.objects.NativeMath
import org.joml.Vector3f
import org.recast4j.Vectors
import org.recast4j.detour.NavMeshBuilder.createNavMeshData
import org.recast4j.detour.NavMeshBuilder.subdivide

class MeshData : MeshHeader() {

    val header get() = this

    /**
     * The tile vertices. [Size: MeshHeader::vertCount]
     */
    lateinit var vertices: FloatArray

    /**
     * The tile polygons. [Size: MeshHeader::polyCount]
     */
    lateinit var polygons: Array<Poly>

    /**
     * The tile's detail sub-meshes. [Size: MeshHeader::detailMeshCount]
     */
    var detailMeshes: Array<PolyDetail>? = null

    /**
     * The detail mesh's unique vertices. [(x, y, z) * MeshHeader::detailVertCount]
     */
    lateinit var detailVertices: FloatArray

    /**
     * The detail mesh's triangles. [(vertA, vertB, vertC) * MeshHeader::detailTriCount] See DetailTriEdgeFlags and
     * NavMesh::getDetailTriEdgeFlags. Unsigned bytes; 4 per each detail triangle (???)
     */
    lateinit var detailTriangles: ByteArray

    /**
     * The tile bounding volume nodes. [Size: MeshHeader::bvNodeCount] (Will be null if bounding volumes are disabled.)
     */
    var bvTree: Array<BVNode>? = null

    /**
     * The tile off-mesh connections. [Size: MeshHeader::offMeshConCount]
     */
    lateinit var offMeshCons: Array<OffMeshConnection>

    companion object {
        fun build(params: NavMeshDataCreateParams, tileX: Int, tileY: Int): MeshData? {
            val data = createNavMeshData(params)
            if (data != null) {
                data.x = tileX
                data.y = tileY
            }
            return data
        }

        fun build(data: MeshData) {
            data.bvTree = Array(data.polyCount * 2) { BVNode() }
            data.bvNodeCount =
                if (data.bvTree!!.isEmpty()) 0 else
                    createBVTree(data, data.bvTree!!, data.bvQuantizationFactor)
        }

        private fun createBVTree(data: MeshData, BVNodes: Array<BVNode>, quantFactor: Float): Int {
            val items = Array(data.polyCount) { BVNode() }
            for (i in 0 until data.polyCount) {
                val node = items[i]
                node.index = i
                val bmin = Vector3f()
                bmin.set(data.vertices, data.polygons[i].vertices[0] * 3)
                val bmax = Vector3f(bmin)
                for (j in 1 until data.polygons[i].vertCount) {
                    val idx = data.polygons[i].vertices[j] * 3
                    NativeMath.min(bmin, data.vertices, idx)
                    NativeMath.max(bmax, data.vertices, idx)
                }
                val header = data
                val headerMin = header.bmin
                node.minX = quantitize(bmin.x, headerMin.x, quantFactor)
                node.minY = quantitize(bmin.y, headerMin.y, quantFactor)
                node.minZ = quantitize(bmin.z, headerMin.z, quantFactor)
                node.maxX = quantitize(bmax.x, headerMin.x, quantFactor)
                node.maxY = quantitize(bmax.y, headerMin.y, quantFactor)
                node.maxZ = quantitize(bmax.z, headerMin.z, quantFactor)
            }
            return subdivide(items, 0, data.polyCount, 0, BVNodes)
        }

        private fun quantitize(bmin: Float, headerMin: Float, quantFactor: Float): Int {
            return Vectors.clamp(((bmin - headerMin) * quantFactor).toInt(), 0, 0x7fffffff)
        }
    }
}